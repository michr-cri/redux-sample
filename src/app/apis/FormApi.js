import 'jquery';

class FormApi {
    static fetchSeedData() {
        return jQuery.ajax({
            method: 'GET',
            url: APP_URL + '/states-seed',
            header: {
                'Content-Type': 'application/json'
            }
        });
    }
    static fetchInitialData() {
        return jQuery.ajax({
            method: 'GET',
            url: APP_URL + '/states',
            header: {
                'Content-Type': 'application/json'
            }
        });
    }

    static saveFormData(selectedItem) {
        return jQuery.ajax({
            method: 'POST',
            url: APP_URL + '/states',
            data: selectedItem,
            header: {
                'Content-Type': 'application/json'
            }
        });
    }
}

export default FormApi;