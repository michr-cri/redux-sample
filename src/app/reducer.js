import { combineReducers } from 'redux'
import typeioReducer from './state/reducers/reducer.typeio';
import feedbackReducer from './shared/reducers/reducer.feedback';
import stateReducer from './state/reducers/reducer.state';
import loginReducer from './login/reducers/reducer.login';
import cityReducer from './city/reducers/reducer.city';
import {reducer as reduxFormReducer} from 'redux-form';

export default combineReducers({
    state: stateReducer,
    typeio: typeioReducer,
    feedback: feedbackReducer,
    login: loginReducer,
    city: cityReducer,
    form: reduxFormReducer
});