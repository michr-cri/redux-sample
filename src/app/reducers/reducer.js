import { combineReducers } from 'redux'
import typeioReducer from './reducer.typeio';
import feedbackReducer from './reducer.feedback';
import appReducer from './reducer.app';
import loginReducer from './reducer.login';

export default combineReducers({
    app: appReducer,
    typeio: typeioReducer,
    feedback: feedbackReducer,
    login: loginReducer
});