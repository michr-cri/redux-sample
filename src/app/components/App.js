import React from 'react';
import TypeIOContainer from '../container/TypeIOContainer';
import FormApi from '../apis/FormApi';

class App extends React.Component {
    render() {
        return ( <div>
            <ul className='event-list'>
                {this.props.selectedItems.map((selectedItem, index) => <li key={index}>{selectedItem}</li>)}
            </ul>
            <TypeIOContainer />
            <div>
                <button id='buttonAdd'
                        onClick={()=> {
                            this.props.addSelectedItems.call(this, this.props.newSelectedItems);
                        }}>
                    Add
                </button>
            </div>
            <div>
                <button id='buttonSave'
                        onClick={()=> {
                            this.props.saveSelectedItems.call(this, this.props.selectedItems);
                        }}>
                    Save
                </button>
            </div>
        </div> );
    }
}

export default App;
