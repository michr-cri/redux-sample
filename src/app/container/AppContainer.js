import { connect } from 'react-redux';
import App from '../components/App';

function mapStateToProps(state) {
    return {
        selectedItems: state.app.get('selectedItems'),
        newSelectedItems: state.typeio.get('selectedItems')
    }
}

function mapDispatchToProps(dispatch) {
    return {
        addSelectedItems: (newSelectedItems) => {
            dispatch({type: 'ADD_SELECTED_ITEMS', payload: newSelectedItems});
        }
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(App);