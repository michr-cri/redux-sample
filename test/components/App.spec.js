import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import ReactTestUtils from 'react-dom/test-utils';
import {findDOMNode} from 'react-dom';
import { Provider } from 'react-redux';

import App from '../../src/app/components/App';
import AppContainer from '../../src/app/container/AppContainer';
import reducer from '../../src/app/reducers/reducer';
import {createStore} from 'redux';
import {Router} from 'react-router-dom';
import history from '../../src/app/shared/history';
import FormApi from '../../src/app/apis/FormApi';

import jQuery from 'jquery';
window.$ = window.jQuery = jQuery;

let app;

describe('components', () => {
    describe('App', () => {
        beforeEach(()=> {
            let source = [{text: 'Michigan', value: 'MI'}, {text: 'New York', value: 'NY'}];
            let initialResults = [{text: 'Michigan', value: 'MI'}];
            let fetchSeedDataPromise = new Promise(function(resolve) {resolve(source);});
            let fetchInitialDataPromise = new Promise(function(resolve) {resolve(initialResults);});

            jest.spyOn(FormApi, 'fetchSeedData').mockReturnValueOnce(fetchSeedDataPromise);
            jest.spyOn(FormApi, 'fetchInitialData').mockReturnValueOnce(fetchInitialDataPromise);
            let store = createStore(reducer);
            app = ReactTestUtils.renderIntoDocument(
                <Provider store={store}>
                    <Router history={history} >
                        <AppContainer />
                    </Router>
                </Provider>
            );
        });

        it('should render self', () => {
            const props = {
                selectedItems: ['item1', 'item2'],
                addSeletedItems: jest.fn()
            };
            const renderer = new ReactShallowRenderer();
            const component = renderer.render(<App {...props} />);

            expect(component).toMatchSnapshot();
        });

        it('should add one item when add button is clicked', () => {
            app.forceUpdate();

            const appDOM = findDOMNode(app);

            let eventsLength = appDOM.querySelectorAll('.event-list>li').length;

            let addInput = appDOM.querySelector('#exampleInput');
            $(addInput).selectItem('New York');

            let addButton = appDOM.querySelector('button');
            ReactTestUtils.Simulate.click(addButton);

            expect(appDOM.querySelectorAll('.event-list>li').length).toEqual(eventsLength + 1);

            expect(appDOM.querySelectorAll('.event-list>li')[0].innerHTML).toContain('MI');
            expect(appDOM.querySelectorAll('.event-list>li')[1].innerHTML).toContain('NY');
        });
    })
});

