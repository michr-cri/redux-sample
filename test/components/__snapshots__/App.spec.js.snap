// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components App should render self 1`] = `
<div>
  <ul
    className="event-list"
  >
    <li>
      item1
    </li>
    <li>
      item2
    </li>
  </ul>
  <Connect(TypeIO) />
  <div>
    <button
      id="buttonAdd"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <div>
    <button
      id="buttonSave"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;
